kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/internal/filter/NettyFilterDispatchExecutorTest.java

```diff
@@ -0,0 +1,347 @@
+            CompletableFuture<Object> completionStage = new CompletableFuture<>();
+            CompletionStage<Boolean> inDispatchThread = dispatchExecutor.completeOnFilterDispatchThread(completionStage)
+                    .thenApply(o -> dispatchExecutor.isInFilterDispatchThread());
+            completionStage.complete("arbitrary");
+            assertThat(inDispatchThread).succeedsWithin(5, TimeUnit.SECONDS).isEqualTo(true);
+        }
+        finally {
+            eventLoop.shutdownGracefully();
+        }
+    }
+
+    // sanity check, future is completed as expected when the input future is completed by the dispatcher thread
+    @Test
+    void completeOnFilterDispatchThreadWhenInDispatchThread() {
+        EventLoop eventLoop = new DefaultEventLoop();
+        try {
+            NettyFilterDispatchExecutor dispatchExecutor = new NettyFilterDispatchExecutor(eventLoop);
+            CompletableFuture<Object> completionStage = new CompletableFuture<>();
+            CompletionStage<Boolean> inDispatchThread = dispatchExecutor.completeOnFilterDispatchThread(completionStage)
+                    .thenApply(o -> dispatchExecutor.isInFilterDispatchThread());
+            dispatchExecutor.execute(() -> completionStage.complete("arbitrary"));
+            assertThat(inDispatchThread).succeedsWithin(5, TimeUnit.SECONDS).isEqualTo(true);
+        }
+        finally {
+            eventLoop.shutdownGracefully();
+        }
+    }
+
+    @Test
+    void completeOnFilterDispatchThreadWhenCompletedByAnotherEventLoop() {
+        EventLoop eventLoop = new DefaultEventLoop();
+        EventLoop anotherLoop = new DefaultEventLoop();
+        try {
+            NettyFilterDispatchExecutor dispatchExecutor = new NettyFilterDispatchExecutor(eventLoop);
+            CompletableFuture<Object> completionStage = new CompletableFuture<>();
+            CompletionStage<Boolean> inDispatchThread = dispatchExecutor.completeOnFilterDispatchThread(completionStage)
+                    .thenApply(o -> dispatchExecutor.isInFilterDispatchThread());
+            anotherLoop.execute(() -> completionStage.complete("arbitrary"));
+            assertThat(inDispatchThread).succeedsWithin(5, TimeUnit.SECONDS).isEqualTo(true);
+            // double-check we are on the expected eventloop
+            CompletableFuture<Object> completionStage2 = new CompletableFuture<>();
+            CompletionStage<Boolean> inExpectedEventLoop = dispatchExecutor.completeOnFilterDispatchThread(completionStage2)
+                    .thenApply(o -> eventLoop.inEventLoop() && !anotherLoop.inEventLoop());
+            anotherLoop.execute(() -> completionStage2.complete("arbitrary"));
+            assertThat(inExpectedEventLoop).succeedsWithin(5, TimeUnit.SECONDS).isEqualTo(true);
+        }
+        finally {
+            eventLoop.shutdownGracefully();
+            anotherLoop.shutdownGracefully();
+        }
+    }
+
```

Comment:
add completeOnFilterDispatchThread tests that verify behaviour when a caller tried to complete or cancel the stage returned by completeOnFilterDispatchThread.

Reply:
Have added tests checking that it can't be converted or cast to a `CompletableFuture`. `InternalCompletableFuture` prevents conversion using `CompletionStage#toCompletableFuture()`. You can't get at something to complete/cancel.